src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,920 @@
+                }
+                else
+                {
+                    return SyntaxFactory.Block(newStatement, statement);
+                }
+            }
+
+            private bool IsLocalOrParameterSymbol(IOperation operation)
+            {
+                if (operation is IConversionOperation conversion && conversion.IsImplicit)
+                {
+                    return IsLocalOrParameterSymbol(conversion.Operand);
+                }
+
+                return operation.Kind == OperationKind.LocalReference || operation.Kind == OperationKind.ParameterReference;
+            }
+
+            private static BlockSyntax WrapWithBlock(StatementSyntax statement)
+                => statement is BlockSyntax block ? block : SyntaxFactory.Block(statement);
+
+            // Checks if the node is within an immediate lambda or within an immediate anonymous method.
+            // 'lambda => node' returns true
+            // 'lambda => localfunction => node' returns false
+            // 'member => node' returns false
+            private static bool IsWithinImmediateLambdaOrAnonymousMethod(SyntaxNode node)
+            {
+                while (node != null)
+                {
+                    switch (node.Kind())
+                    {
+                        case SyntaxKind.AnonymousMethodExpression:
+                        case SyntaxKind.ParenthesizedLambdaExpression:
+                        case SyntaxKind.SimpleLambdaExpression:
+                            return true;
+                        case SyntaxKind.LocalFunctionStatement:
+                            return false;
+                        default:
+                            if (node is MemberDeclarationSyntax)
+                            {
+                                return false;
+                            }
+
+                            break;
+                    }
+
+                    node = node.Parent;
+                }
+
+                return false;
+            }
+
+            private class QueryExpressionProcessingInfo
+            {
+                public Stack<CSharpSyntaxNode> Stack { get; private set; }
+
+                public List<SyntaxToken> Identifiers { get; private set; }
+
+                public QueryExpressionProcessingInfo(FromClauseSyntax fromClause)
+                {
+                    Stack = new Stack<CSharpSyntaxNode>();
+                    Stack.Push(fromClause);
+                    Identifiers = new List<SyntaxToken>();
+                    Identifiers.Add((fromClause.Identifier));
+                }
+
+                public bool TryAdd(CSharpSyntaxNode node, SyntaxToken identifier)
+                {
+                    // Duplicate identifiers are not allowed.
+                    // var q = from x in new[] { 1 } select x + 2 into x where x > 0 select 7 into y let x = ""aaa"" select x;
+                    if (ContainsIdentifier(identifier))
```

Comment:
>if (ContainsIdentifier(identifier)) [](start = 20, length = 35)

If ```Identifiers``` was a HashSet, this could be replaced with ```if(!Identifiers.Add(identifier.ValueText))```.